pytential/symbolic/mappers.py

```diff
@@ -137,6 +139,66 @@ def map_interpolation(self, expr):
 
         return type(expr)(expr.from_dd, expr.to_dd, operand)
 
+# }}}
+
+
+# {{{ IntGTermCollector
+
+class IntGTermCollector(IdentityMapper):
+    r"""A mapper that removes all non-:class:`~pytential.symbolic.primitves.IntG`
+    terms from the expression and all their non-constant factors.
+
+    In particular, an expression of the type
+
+    .. math::
+
+        \sum_{i = 0}^N f_i(\mathbf{x}, \sigma)
+        + \sum_{i = 0}^M c_i g_i(\mathbf{x}) \mathrm{IntG}_i(\mathbf{x})
+
+    is reduced to
+
+    .. math::
+
+        \sum_{i = 0}^M c_i \mathrm{IntG}_i(\mathbf{x}).
+
+    The intended used of this transformation is to allow the evaluation of
+    the proxy interactions in :ref:`_direct_solver` for a given expression
+    meant for self-evaluation.
+    """
+
+    def map_sum(self, expr):
+        collector = OperatorCollector()
+
+        children = []
+        for child in expr.children:
+            rec_child = self.rec(child)
+            if collector(rec_child):
+                children.append(rec_child)
+
+        from pymbolic.primitives import flattened_sum
+        return flattened_sum(children)
+
+    def map_product(self, expr):
+
+        collector = OperatorCollector()
+
+        from pymbolic.primitives import is_constant
+        children = []
+        for child in expr.children:
+            rec_child = self.rec(child)
+            if is_constant(child) or collector(rec_child):
+                children.append(rec_child)
```

Comment:
Maybe complain if there are more than one of these? Otherwise this is unlikely to be a linear operator...

Reply:
Yeah, good point! Changed in cf1d364.